clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -1279,4 +1288,119 @@ static class MemberInfo {
             this.memberEpoch = Optional.empty();
         }
     }
+
+    /**
+     * As the name suggests, this caches the result of an {@link OffsetFetchRequestState offset fetch request}
+     * <em>in specific cases</em>. The logic resembles that of {@link ConsumerCoordinator}'s
+     * <code>PendingCommittedOffsetRequest</code> mechanism.
+     *
+     * <p/>
+     *
+     * This handles the case where a user calls {@link Consumer#poll(Duration)} with a low timeout value, such as 0.
+     * As the timeout value approaches zero, the likelihood that the client will be able to fetch the offsets from the
+     * broker within the user's timeout also decreases. But we can take advantage of the fact that {@code poll()}
+     * is typically invoked in a tight loop, so the user's application will likely call {@code poll()} again to make
+     * a second attempt with the same set of {@link TopicPartition}s as the first attempt.
+     *
+     * <p/>
+     *
+     * The idea here is to cache the results of the last successful--but expired--response. An operation may exceed
+     * the time the user allotted, but that doesn't mean that the network request is aborted. In this scenario, it is
+     * often the case that the client receives a successful response, though it has technically exceeded the amount
+     * of time the user specified. However, as mentioned above, {@code poll()} is likely to be invoked again with the
+     * same set of partitions. By caching the successful response from attempt #1--though it timed out from the user's
+     * perspective--the client is able to satisfy attempt #2 from the cache.
```

Comment:
What is the expected behavior if there is a commit between the two calls to poll()?
If the first offset fetch returns successfully before the commit is done, the returned offsets would be wrong, wouldn't they?

Reply:
That's a very good point. The legacy logic does not cache the fetch responses. It caches only the fetch request for which it hasn't received a response yet (that can be safely done if the set of partitions requested are the same). As soon as it gets a response it clears the cached request ([here](https://github.com/apache/kafka/blob/f746d67c3bf62e7d4f5f4e652ea36b4d5a0e01a6/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L947)).  We should never cache the fetch responses because it could definitely lead to the gap that @cadonna is pointing out, if a commit happens in between.

Reply:
@lianetm in the legacy consumer, does the fact that the offset fetch request and the offset commit request use the same network connection ensure that the requests have an order and the commit request cannot overtake the fetch request if commit was sent after fetch? Sorry if this question seems dumb, but I am not familiar with the network part.

Reply:
I believe so, but mainly because I expect that the broker will handle them one at a time, so it will receive a fetch first, a commit after, and it won't handle the commit until it's done with the previous fetch request. So legacy and new logic, it's all safe I would say until the moment we start caching on the client side. There we have to be very careful not to return cached fetch results (that we may have received before a commit).

Reply:
I've added a call to clear the cache whenever a commit (of any sort) is enqueued.